setuptools/pep517.py

```diff
@@ -0,0 +1,91 @@
+import os
+import sys
+import subprocess
+import tokenize
+import shutil
+import tempfile
+
+from setuptools import dist
+from setuptools.dist import SetupRequirementsError
+
+
+SETUPTOOLS_IMPLEMENTATION_REVISION = 0.1
+
+def _run_setup(setup_script='setup.py'): #
+    # Note that we can reuse our build directory between calls
+    # Correctness comes first, then optimization later    
+    __file__=setup_script
+    f=getattr(tokenize, 'open', open)(__file__)
+    code=f.read().replace('\\r\\n', '\\n')
+    f.close()
+    exec(compile(code, __file__, 'exec'))
+
+
+def fix_config(config_settings):
+    config_settings = config_settings or {}
+    config_settings.setdefault('--global-option', [])
+    return config_settings
+
+def get_build_requires(config_settings):
+    config_settings = fix_config(config_settings)
+    requirements = ['setuptools', 'wheel']
+    dist._skip_install_eggs = True
+
+    sys.argv = sys.argv[:1] + ['egg_info'] + \
+        config_settings["--global-option"]
+    try:
+        _run_setup()
+    except SetupRequirementsError as e:
+        requirements += e.specifiers
+
+    dist._skip_install_eggs = False
+
+    return requirements
+
+
+def get_requires_for_build_wheel(config_settings=None):
+    config_settings = fix_config(config_settings)
+    return get_build_requires(config_settings)
+
+
+def get_requires_for_build_sdist(config_settings=None):
+    config_settings = fix_config(config_settings)
+    return get_build_requires(config_settings)
+
```

Comment:
These two `get_requires` functions are the public API. They will be used by pip to determine the setup requirements.